tls/s2n_client_hello.c

```diff
@@ -241,11 +241,16 @@ static int s2n_parse_client_hello(struct s2n_connection *conn)
     POSIX_GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
     POSIX_ENSURE_REF(ecc_pref);
 
-    /* This is going to be our fallback if the client has no preference. */
-    /* A TLS-compliant application MUST support key exchange with secp256r1 (NIST P-256) */
-    /* and SHOULD support key exchange with X25519 [RFC7748]. */
-    /* - https://tools.ietf.org/html/rfc8446#section-9.1 */
-    conn->secure.server_ecc_evp_params.negotiated_curve = &s2n_ecc_curve_secp256r1;
+    if (s2n_ecc_preferences_includes_curve(ecc_pref, TLS_EC_CURVE_SECP_256_R1)) {
+        /* This is going to be our fallback if the client has no preference. */
+        /* A TLS-compliant application MUST support key exchange with secp256r1 (NIST P-256) */
+        /* and SHOULD support key exchange with X25519 [RFC7748]. */
+        /* - https://tools.ietf.org/html/rfc8446#section-9.1 */
+        conn->secure.server_ecc_evp_params.negotiated_curve = &s2n_ecc_curve_secp256r1;
+    } else {
+        /* P-256 is the preferred fallback option. These prefs don't support it, so choose whatever curve is first. */
+        conn->secure.server_ecc_evp_params.negotiated_curve = ecc_pref->ecc_curves[0];
```

Comment:
Is it guaranteed that we always have at least one curve?

Reply:
I think we need a defensive check. Can add one to the conditional here but also thinking of other places to have the check to short circuit this logic sooner.

Reply:
Added a sanity check during s2n_security_policies_init() where @zz85's previous p256 check was.

Reply:
Maybe `POSIX_ENSURE_GT(ecc_pref->count,  0);` for extra paranoia? (:

Reply:
will do!